tinygrad/codegen/linearizer.py

```diff
@@ -410,28 +411,33 @@ def linearize(self):
     self.applied_opts_cache = self.applied_opts[:]
     return self
 
-  def ast_parse(self, x:LazyOp, acc: List[UOp], offs:Optional[List[int]], loaded_buffers:Dict[Union[MemBuffer, ConstBuffer, LocalBuffer], List[UOp]], do_reduce=False, loop_ctx=tuple(), cache=None) -> List[UOp]: # noqa: E501
+  def ast_parse(self, x:LazyOp, acc: List[UOp], offs:Optional[List[int]], loaded_buffers:Dict[Union[MemBuffer, ConstBuffer, LocalBuffer], List[UOp]], \
+                do_reduce=False, loop_ctx=tuple(), if_ctx=tuple(), insert_before:Optional[UOp]=None, cache=None) -> List[UOp]: # noqa: E501
     if cache is None: cache = {}
     if x in cache: return cache[x]
     if x.op in BufferOps: return loaded_buffers[x.arg]
-    if x.op is UnaryOps.CAST: return [self.uops.add(UOps.BITCAST if x.arg[1] else UOps.CAST, self.get_base_dtype(x.arg[0]), (u,), x.arg[0]) \
-                                      for u in self.ast_parse(x.src[0], acc, offs, loaded_buffers)]
+    if x.op is UnaryOps.CAST: return [self.uops.add(UOps.BITCAST if x.arg[1] else UOps.CAST, self.get_base_dtype(x.arg[0]), (u,), x.arg[0], insert_before=insert_before) \
+                                      for u in self.ast_parse(x.src[0], acc, offs, loaded_buffers, insert_before=insert_before)]  # noqa: E501
     if x.op in ReduceOps and not do_reduce:
-      assert offs is None, "not available if we aren't doing reduce"
-      return acc
-
-    values = [self.ast_parse(v, acc, offs, loaded_buffers, loop_ctx=loop_ctx, cache=cache) for v in x.src]
+      if x not in self.reduce_acc:
+        load_cache_copy,self.load_cache = self.load_cache,{}
```

Comment:
I don't understand this change

Reply:
yes and no

if control flow reaches here it means `ast_parse` has run into a reduceop that hasn't been rendered, so it needs to go back before any loop or if statements, render the reduceop, then it can continue with the parse

with only one reduceop I don't think it's safe to keep `self.load_cache` because I'm not sure you ever run into a case where you have to leave a context

but with >1 reduceop you will, and so you'll get a UOp graph like below, where the second reduceop will try to reference a future load that was in `self.load_cache`
0: `ast_parse` hits this reduceop first
1: `render_reduceop` loads the value it needs
2: `ast_parse` (from within `render_reduceop`) runs into another reduceop
3: renders it here
4: `ast_parse` tries to reference a future load because it's in `self.load_cache`
```
   0 UOps.DEFINE_GLOBAL  : ptr.dtypes.float          []                               (0, 'data0', True)
   1 UOps.DEFINE_GLOBAL  : ptr.dtypes.float          []                               (1, 'data1', False)
   2 UOps.SPECIAL        : dtypes.int                []                               (0, 'gidx0', 128)
   3 UOps.DEFINE_ACC     : dtypes.float              []                               0.0
   4 UOps.CONST          : dtypes.int                []                               0
   5 UOps.CONST          : dtypes.int                []                               64
   6 UOps.ALU            : dtypes.int                [2, 5]                           BinaryOps.MUL
   7 UOps.LOOP           : dtypes.int                [4, 5]                           None                <- 3
   8 UOps.ALU            : dtypes.float              [13, 3]                          BinaryOps.ADD       <- 4
   9 UOps.PHI            : dtypes.float              [3, 8, 7]                        None
  10 UOps.ENDLOOP        :                           [7]                              None
  11 UOps.LOOP           : dtypes.int                [4, 5]                           None                <- 0
  12 UOps.ALU            : dtypes.int                [6, 11]                          BinaryOps.ADD
  13 UOps.LOAD           : dtypes.float              [1, 12]                          None                <- 1
  14 UOps.ALU            : dtypes.float              [13, 9]                          BinaryOps.SUB       <- 2
  15 UOps.ALU            : dtypes.float              [14, 3]                          BinaryOps.ADD
  16 UOps.PHI            : dtypes.float              [3, 15, 11]                      None
  17 UOps.ENDLOOP        :                           [11]                             None
  18 UOps.STORE          :                           [0, 2, 16]                       None
```

Reply:
I think copying the load cache hides the problem instead of solving the root cause.